workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs

```diff
@@ -40,18 +48,52 @@ protected override void OnUpdate()
 
         private void OnAddEntity(AddEntityOp op)
         {
-            worker.CreateEntity(op.EntityId.Id);
+            var entityId = op.EntityId.Id;
+            if (worker.EntityMapping.ContainsKey(entityId))
+            {
+                worker.LogDispatcher.HandleLog(LogType.Error, new LogEvent("Tried to add an entity but there is already an entity associated with that EntityId.")
+                    .WithField(LoggingUtils.LoggerName, LoggerName)
+                    .WithField(LoggingUtils.EntityId, entityId));
+                return;
+            }
+
+            var entity = EntityManager.CreateEntity();
+            EntityManager.AddComponentData(entity, new SpatialEntityId
+            {
+                EntityId = entityId
+            });
+            EntityManager.AddComponentData(entity, new NewlyAddedSpatialOSEntity());
+
+            translationUnityRegistry.AddAllCommandRequestSenders(entity, entityId);
+            worker.EntityMapping.Add(entityId, entity);
         }
 
         private void OnRemoveEntity(RemoveEntityOp op)
         {
-            worker.RemoveEntity(op.EntityId.Id);
+            var entityId = op.EntityId.Id;
+            if (!worker.TryGetEntity(entityId, out var entity))
+            {
+                worker.LogDispatcher.HandleLog(LogType.Error, new LogEvent("Tried to delete an entity but there is no entity associated with that EntityId.")
```

Comment:
Duplicate error strings, could extract into a separate variable.

Reply:
Yup - magic strings should be const